src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/CardinalityBasedJoinPlanOptimizerBase.java

```diff
@@ -0,0 +1,262 @@
+                final int numOfSubPlans = plan.numberOfSubPlans();
+                for (int i = 0; i < numOfSubPlans; i++) {
+                    reqOps.addAll( extractAllRequestOpsFromSourceAssignment(plan.getSubPlan(i)) );
+                }
+                return reqOps;
+            }
+            else
+                throw new IllegalArgumentException("Unsupported type of subquery in source assignment (" + pop.getClass().getName() + ")");
+        }
+
+        /**
+         *
+         * @param reqOpsOfAllSubPlans A list of request operators, which is ordered in the order in which the request operators can be found by a depth-first traversal of the subplans.
+         * @param resps A list of cardinality in the same order of 'reqOpsOfAllSubPlans'
+         * @return A list of PhysicalPlanWithStatistics, with each object corresponding to a subplan in the 'subplans' list.
+         * Each PhysicalPlanWithStatistics object contains the physical plan for the subplan, along with its cardinality,
+         * a list of relevant federations members, and estimated number of access to federation members.
+         *
+         * During a depth-first traversal of the subplans, increase the index by 1 when visiting a request operator:
+         * If the root operator of the subplan is a Request operator, get the corresponding cardinality from resps with the same index;
+         * If the subPlan with a Union of requests, calculate the total cardinality by summing up the cardinality of individual requests.
+         * (See page 1052 in the paper[1])
+         *
+         * [1] Heling, Lars, and Maribel Acosta. "Federated SPARQL query processing over heterogeneous linked data fragments." Proceedings of the ACM Web Conference 2022.
+         */
+        protected List<PhysicalPlanWithStatistics> associateCardWithSubPlans(
+                final List<PhysicalPlan> subplans,
+                final List<LogicalOpRequest<?,?>> reqOpsOfAllSubPlans,
+                final CardinalityResponse[] resps ){
+            final List<PhysicalPlanWithStatistics> subPlansWithStatistics = new ArrayList<>();
+
+            int index = 0;
+            for ( final PhysicalPlan subplan : subplans ) {
+                final PhysicalOperator pop = subplan.getRootOperator();
+
+                PhysicalPlanWithStatistics planWithStatistics = null;
+                if ( pop instanceof PhysicalOpRequest
+                        || (pop instanceof PhysicalOpFilter && subplan.getSubPlan(0).getRootOperator() instanceof PhysicalOpRequest)) {
+                    final int cardinality = resps[index].getCardinality();
+                    final FederationMember fm = reqOpsOfAllSubPlans.get(index).getFederationMember();
+                    final int numOfAccess = accessNumForReq(cardinality, fm);
+
+                    planWithStatistics = new PhysicalPlanWithStatistics( subplan, Arrays.asList(fm), null, cardinality, numOfAccess );
```

Comment:
Shouldn't the vocab.mapping of the fed.member (if any) be passed to the constructor as well?